src/iface/ethernet.rs

```diff
@@ -1049,34 +1063,38 @@ mod test {
         // Ensure the ident we bound to and the ident of the packet are the same.
         let mut bytes = [0xff; 24];
         let mut packet = Icmpv4Packet::new(&mut bytes);
-        let echo_repr = Icmpv4Repr::EchoReply {
-            ident:  ident,
-            seq_no: 0x5432,
-            data:   &[0xff; 16],
-        };
+        let echo_repr = Icmpv4Repr::EchoRequest{ ident, seq_no, data: echo_data };
         echo_repr.emit(&mut packet, &ChecksumCapabilities::default());
-        let data = &packet.into_inner()[..];
+        let icmp_data = &packet.into_inner()[..];
 
-        let ip_repr = IpRepr::Ipv4(Ipv4Repr {
+        let ipv4_repr = Ipv4Repr {
             src_addr:    Ipv4Address::new(0x7f, 0x00, 0x00, 0x02),
             dst_addr:    Ipv4Address::new(0x7f, 0x00, 0x00, 0x01),
             protocol:    IpProtocol::Icmp,
             payload_len: 24,
             ttl:         64
-        });
+        };
+        let ip_repr = IpRepr::Ipv4(ipv4_repr);
 
         // Open a socket and ensure the packet is handled due to the listening
         // socket.
         {
             assert!(!socket_set.get::<IcmpSocket>(socket_handle).can_recv());
         }
-        assert_eq!(iface.inner.process_icmpv4(&mut socket_set, ip_repr, data),
-                   Ok(Packet::None));
+
+        // Confirm we still get EchoReply from `smoltcp` even with the ICMP socket listening
+        let echo_reply = Icmpv4Repr::EchoReply{ ident, seq_no, data: echo_data };
+        let mut ipv4_reply = ipv4_repr.clone();
+        mem::swap(&mut ipv4_reply.src_addr, &mut ipv4_reply.dst_addr);
+        assert_eq!(iface.inner.process_icmpv4(&mut socket_set, ip_repr, icmp_data),
+                   Ok(Packet::Icmpv4((ipv4_reply, echo_reply))));
+
         {
             let mut socket = socket_set.get::<IcmpSocket>(socket_handle);
             assert!(socket.can_recv());
             assert_eq!(socket.recv(),
-                       Ok((&data[..], IpAddress::Ipv4(Ipv4Address::new(0x7f, 0x00, 0x00, 0x02)))));
+                       Ok((&icmp_data[..],
+                           IpAddress::Ipv4(Ipv4Address::new(0x7f, 0x00, 0x00, 0x02)))));
```

Comment:
This IP is used multiple times. Might be worthwhile to make this a variable that is used.